### ISSUE TITLE
syntax priority problem

### ISSUE THREAD

**Issue by grosu** [kframework/k-legacy#184]
```
module TRY
  syntax Exp  ::= Int
                | Exp Exp
                | Exp "-" Exp
                | "-" Exp

  syntax priority '__ > '_-_
endmodule

fslwork$ krun --parser "kast -e" "7 - 3"
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: Exp ::= Exp "-" Exp 
   7 - 3
2: Exp ::= Exp Exp 
   7 (- 3)
    File: 7 - 3
    Location: (1,1,1,6)
    Compilation Phase: Ambiguity filter
```

There should be no ambiguity, because application has priority over subtraction.

**Comment by grosu** [comment]
Just to make it clear: I know why this happens, and I know that Radu will say this is not a bug, but I disagree.  Maybe we need a different filter here, something like a "targeted prefer", e.g.

```
syntax prefer '__ > '_-_
```

 but we need something for such cases.

**Comment by radumereuta** [comment]
But you do have the means to disambiguate this grammar.
1) prefer/avoid attributes. Unfortunately they are non-discriminatory therefore should be avoided if possible (not the case here)
2) What is the relation between the unary minus and the other productions? You didn't specify anything like that in your grammar.

Again, what priority means is that it filters parses where A is a parent of B and you have the relation A > B.
The classic example is \* and +
   Exp "_" Exp > Exp "+" Exp
For the input "1 + 2 \* 3" you will get the following ambiguous trees:
amb('_+_(1, '___(2, 3)),
       '_*_('_+_(1, 2), 3))

The valid parse in this case is the first one. The second one is rejected because \* is a parent of +.

We had a long discussion at one point regarding prefer versus priority and we concluded that prefer is really not enough. The problem appears when you have something like "1 + 2 + 3 \* 4"... I won't start this discussion again, at least not here.

**Comment by grosu** [comment]
Of course, we know how the priority filter works.  That was not the message of my bug report.  The idea is that we want a discriminatory prefer filter as well.  I therefore change the tag of this issue from bug to enhancement.  Since we are going to change the front end parser anyway, we should not worry about performance too much at this stage.  I would be more worried about having the right functionality so we can define any syntax we want.

**Comment by radumereuta** [comment]
I've found the previous email thread that we had one year ago, and I've resent it on the k-list. New people joined the group since then, so they would be particularly interested in reading that.

We should continue this discussion with the lessons we learned from there.

**Comment by adamsmd** [comment]
I don't see how this is a bug at all as the priority declaration that you gave is wrong.  It should be "syntax priority '__ > '-_".

That also has the side effect of making "-A B" parse as "-(A B)" instead of "(-A) B".  If you want the latter, then we need to support the "positional" priorities that we've already discussed multiple times and plan to add with the new parser.

**Comment by grosu** [comment]
Yes, that solves that particular problem, and so does the global prefer.  The more I use it, the more I find the word "priority" confusing.  Take our example: "application has higher priority than unary minus, so 7-3 parses as binary minus".  Also, however we end up calling it, in my view it fails to be modular: we need to define certain relationships between _certain_ productions, in order for a string to parse as _another_ production.  Besides, such "priorities" or whatever they are should only apply when we have ambiguities, not always.  For example, if I remove the _-_ production, I expect 7-3 to parse correctly, but we get an error:

```
fslwork$ krun --parser "kast -e" "7 - 3"
[Error] Critical: Parse error: Syntax error near unexpected character ' '
    File: 7 - 3
    Location: (1,4,1,4)
```

I would really prefer to say something like this:

```
syntax prefer '__ > '_-_
```

Or even better and cleaner:

rule A-B, (A[-B] => .)

**Comment by adamsmd** [comment]
```
fslwork$ krun --parser "kast -e" "7 - 3"
[Error] Critical: Parse error: Syntax error near unexpected character ' '
    File: 7 - 3
    Location: (1,4,1,4)
```

I don't understand the parser is rejecting that.  Could you clarify what is happening here?

**Comment by grosu** [comment]
Why don't you try it.  If you remove the binary minus production, there is only one way to parse the above, namely 7 (- 3).  Yet, it is rejected, because the priority '__ > '-_ applies even when there is no ambiguity at all.

**Comment by adamsmd** [comment]
I never questioned _that_ it is rejected.  What I didn't understand is _why_ it is rejected which "trying it" wouldn't have answered.

However, given the explanation in the rest of your message I now see that when you said "remove the - production", you meant remove the `_-_` production but I assumed you meant the `-_` production (wiki syntax probably ate the underscores).  In that case, yes, it is clear why the parse would be rejected and I agree that using something like `syntax prefer '_-_(_,_) over '__(_,-_)` would be better as not only would it not fire if there is no `_-_` production, but we could even issue a warning if a constructor is referred to that does not exist.
